xline/src/client/mod.rs

```diff
@@ -0,0 +1,183 @@
+use std::net::SocketAddr;
+
+use anyhow::{anyhow, Result};
+use curp::{client::Client as CurpClient, cmd::ProposeId};
+use etcd_client::Client as EtcdClient;
+use prost::Message;
+use uuid::Uuid;
+
+use crate::{
+    rpc::{
+        DeleteRangeResponse, PutResponse, RangeResponse, Request, RequestOp, Response, ResponseOp,
+    },
+    server::command::{Command, KeyRange},
+};
+
+use kv_types::{PutRequest, RangeRequest};
+
+use self::kv_types::DeleteRangeRequest;
+
+/// covert struct between etcd and curp
+mod convert;
+/// Requests used by Client
+pub mod kv_types;
+
+/// Xline client
+#[allow(missing_debug_implementations)] // EtcdClient doesn't implement Debug
+pub struct Client {
+    /// Name of the client
+    name: String,
+    /// Curp client
+    curp_client: CurpClient<Command>,
+    /// Etcd client
+    etcd_client: EtcdClient,
+    /// Use curp client to send requests when true
+    use_curp_client: bool,
+}
+
+impl Client {
+    /// New `Client`
+    ///
+    /// # Panics
+    ///
+    /// Panics if `EtcdClient::connect` fails.
+    #[inline]
+    pub async fn new(
+        leader_index: usize,
+        all_members: Vec<SocketAddr>,
+        use_curp_client: bool,
+    ) -> Self {
+        let etcd_client = EtcdClient::connect(
+            all_members
+                .iter()
+                .map(ToString::to_string)
+                .collect::<Vec<_>>(),
+            None,
+        )
+        .await
+        .unwrap_or_else(|e| panic!("failed to connect to server: {}", e));
+        let curp_client = CurpClient::new(leader_index, all_members).await;
+        Self {
+            name: String::from("client"),
+            curp_client,
+            etcd_client,
+            use_curp_client,
+        }
+    }
+
+    /// set `use_curp_client`
+    #[inline]
+    pub fn set_use_curp_client(&mut self, use_curp_client: bool) {
+        self.use_curp_client = use_curp_client;
+    }
+
+    /// Generate a new `ProposeId`
+    fn generate_propose_id(&self) -> ProposeId {
+        ProposeId::new(format!("{}-{}", self.name, Uuid::new_v4()))
+    }
+
+    /// parse `ResponseOp` into `Response`
+    fn parse_response_op(response_op: ResponseOp) -> Response {
+        if let Some(response) = response_op.response {
+            response
+        } else {
+            panic!("Receive empty ResponseOp");
```

Comment:
Please don't panic, if any unexpected happens. It's a client used by the others, we'd better not crash users thread.